pkg/cri/resource-manager/control/cpu/cpu.go

```diff
@@ -0,0 +1,253 @@
+	MaxFreq                     uint `json:"maxFreq"`
+	EnergyPerformancePreference uint `json:"energyPerformancePreference"`
+}
+
+// cpuClassAllocations contains the information about how cpus are assigned to
+// classes
+type cpuClassAssignments map[string]utils.IDSet
+
+const (
+	cacheKeyCPUAssignments = "CPUClassAssignments"
+)
+
+var log logger.Logger = logger.NewLogger(CPUController)
+
+// Ccontroller singleton instance.
+var singleton *cpuctl
+
+// GetClasses returns all available CPU classes.
+func GetClasses() map[string]Class {
+	return getCPUController().config.getClasses()
+}
+
+// Assign assigns a set of cpus to a class.
+func Assign(class string, cpus ...int) error {
+	// NOTE: no locking implemented anywhere around -> we don't expect multiple parallel callers
+	return getCPUController().assign(class, cpus...)
+}
+
+// getCPUController returns the (singleton) CPU controller instance.
+func getCPUController() *cpuctl {
+	if singleton == nil {
+		singleton = &cpuctl{}
+		singleton.config = singleton.defaultOptions().(*config)
+	}
+	return singleton
+}
+
+// Start initializes the controller for enforcing decisions.
+func (ctl *cpuctl) Start(cache cache.Cache, client client.Client) error {
+	ctl.cache = cache
+
+	// DEBUG: dump the class assignments we have stored in the cache
+	log.Debug("retrieved cpu class assignments from cache:\n%s", utils.DumpJSON(*ctl.getClassAssignments()))
+
+	if err := ctl.configure(); err != nil {
+		// Just print an error. A config update later on may be valid.
+		log.Error("failed apply initial configuration: %v", err)
+	}
+
+	// TODO: We probably could just remove this and the hooks if they are not used
+	pkgcfg.GetModule(ConfigModuleName).AddNotify(getCPUController().configNotify)
+
+	return nil
+}
+
+// Stop shuts down the controller.
+func (ctl *cpuctl) Stop() {
+}
+
+// PreCreateHook handler for the CPU controller.
+func (ctl *cpuctl) PreCreateHook(c cache.Container) error {
+	return nil
+}
+
+// PreStartHook handler for the CPU controller.
+func (ctl *cpuctl) PreStartHook(c cache.Container) error {
+	return nil
+}
+
+// PostStartHook handler for the CPU controller.
+func (ctl *cpuctl) PostStartHook(c cache.Container) error {
+	return nil
+}
+
+// PostUpdateHook handler for the CPU controller.
+func (ctl *cpuctl) PostUpdateHook(c cache.Container) error {
+	return nil
+}
+
+// PostStopHook handler for the CPU controller.
+func (ctl *cpuctl) PostStopHook(c cache.Container) error {
+	return nil
+}
+
+// UpdateHWConfig handler for the CPU controller.
+func (ctl *cpuctl) UpdateConfig(cache cache.Cache) error {
+	return cache.TraversePendingConfig(Assign)
+}
+
+// assign assigns a cpuset to a class
+func (ctl *cpuctl) assign(class string, cpus ...int) error {
+	if _, ok := ctl.config.Classes[class]; !ok {
+		return fmt.Errorf("non-existent cpu class %q", class)
+	}
+
+	if ok := utils.SetCPUsScalingMinFreq(cpus, (int)(ctl.config.Classes[class].MinFreq)); ok != nil {
```

Comment:
Nitpick: usually a return value of the `error` type has been captured to `err` variable. `ok` has captured booleans.

Reply:
Sure, that can be fixed of course. I can update the PR later today and add other comments if there are any.